racket/collects/racket/file.rkt

```diff
@@ -134,60 +135,112 @@
-         (define str-src
-           (cond [(path? source)
-                  (regexp-replace #rx"^<(.*?)>(?=/)"
-                                  (path->relative-string/library source)
-                                  (lambda (_ s) (string-upcase s)))]
-                 [(string? source) source]
-                 [else #f]))
-         (define str-loc
-           (cond [(and line col) (format "-~a-~a" line col)]
-                 [pos (format "--~a" pos)]
-                 [else ""]))
-         (define combined-str (string-append (or str-src "rkttmp") str-loc))
-         (define sanitized-str (regexp-replace* #rx"[<>:\"/\\|]" combined-str "-"))
-         (define max-len 50) ;; must be even
-         (define not-too-long-str
-           (cond [(< max-len (string-length sanitized-str))
-                  (string-append (substring sanitized-str 0 (- (/ max-len 2) 2))
-                                 "----"
-                                 (substring sanitized-str
-                                            (- (string-length sanitized-str)
-                                               (- (/ max-len 2) 2))))]
-                 [else sanitized-str]))
-         #`(app make-temporary-file/proc
-                #,(string-append not-too-long-str "_~a")))]
+       (infer-template)]
+      [(_ #:copy-from copy-from)
+       (not (keyword? #'copy-from))
+       (infer-template #:copy-from #'copy-from)]
+      [(_ #:base-dir base-dir)
+       (not (keyword? #'base-dir))
+       (infer-template #:base-dir #'base-dir)]
+      [(_ #:base-dir base-dir #:copy-from copy-from)
+       (not (or (keyword? #'base-dir)
+                (keyword? #'copy-from)))
+       (infer-template #:copy-from #'copy-from #:base-dir #'base-dir)]
+      [(_ #:copy-from copy-from #:base-dir base-dir )
+       (not (or (keyword? #'base-dir)
+                (keyword? #'copy-from)))
+       (infer-template #:copy-from #'copy-from #:base-dir #'base-dir)]
       [(_ . whatever)
        #'(app make-temporary-file/proc . whatever)])))
 
-(define make-temporary-file/proc
+(define-values [make-temporary-file/proc
+                make-temporary-directory/proc]
   (let ()
-    (define (make-temporary-file [template "rkttmp~a"] [copy-from #f] [base-dir #f])
+    (define (make-temporary-file [template "rkttmp~a"]
+                                 #:copy-from [_copy-from #f]
+                                 #:base-dir [_base-dir #f]
+                                 [copy-from _copy-from]
+                                 [base-dir _base-dir])
+      (do-make-temporary-file 'make-temporary-file
+                              template copy-from base-dir))
+    (define (make-temporary-directory [template "rkttmp~a"]
+                                      #:base-dir [base-dir #f])
+      (do-make-temporary-file 'make-temporary-directory
+                              template 'directory base-dir))
+    (define (do-make-temporary-file who template copy-from base-dir)
```

Comment:
Instead of abstracting over these two functions from the outside in with a single `do-make-temporary-file` function, I suggest making the implementations of `make-temporary-file` and `make-temporary-directory` separate and pulling out some of the common chunks into helper functions.

Reply:
I'm not sure I'm understanding what you have in mind here. The only actual difference based on `copy-from` is: https://github.com/racket/racket/blob/27e40598bf7187040fb2df463c801c63e5f594e5/racket/collects/racket/file.rkt#L277-L281
Presumably `make-temporary-file` still needs to handle the `'directory` case for compatibility, so I'm not sure what else to split out.